File: lib/csharp/se/lth/control/labcomm2014/RenamingEncoder.cs
namespace se.lth.control.labcomm2014 {

  using System;
  using System.Collections.Generic;

  public class RenamingEncoder : WrappingEncoder {

    private Encoder encoder;
    private RenamingRegistry registry;
    private Func<String,String> rename;
    private Dictionary<SampleDispatcher, SampleDispatcher> alias;

    public RenamingEncoder(Encoder encoder,
                           RenamingRegistry registry,
                           Func<String,String> rename)
    : base(encoder) {
      this.encoder = encoder;
      this.registry = registry;
      this.rename = rename;
      alias = new Dictionary<SampleDispatcher, SampleDispatcher>();
    }

    private SampleDispatcher add(SampleDispatcher identity) {
      SampleDispatcher renamed;
      lock(this) {
        if (! alias.TryGetValue(identity, out renamed)) {
          renamed = registry.add(identity, rename(identity.getName()));
          alias.Add(identity, renamed);
        }
      }
      get(identity);
      return renamed;
    }

    private SampleDispatcher get(SampleDispatcher identity) {
      SampleDispatcher renamed;
      lock(this) {
        alias.TryGetValue(identity, out renamed);
      }
      return renamed;
    }
    public override void register(SampleDispatcher identity) {
      encoder.register(add(identity));
    }

    public override void registerSampleRef(SampleDispatcher identity) {
      encoder.registerSampleRef(add(identity));
    }

    public override void begin(SampleDispatcher identity) {
      base.begin(get(identity));
    }

    public override void end(SampleDispatcher identity) {
      base.end(get(identity));
    }

  }

} 

File: lib/java/se/lth/control/labcomm2014/RenamingEncoder.java
package se.lth.control.labcomm2014;

import java.io.IOException;
import java.util.HashMap;

public class RenamingEncoder extends WrappingEncoder {

  public interface Rename {
    String rename(String name);   
  }

  private RenamingRegistry registry;
  private Rename rename;
  private HashMap<SampleDispatcher, SampleDispatcher> alias;

  public RenamingEncoder(Encoder encoder,
                         RenamingRegistry registry,
                         Rename rename) throws IOException {
    super(encoder);
    this.registry = registry;
    this.rename = rename;
    alias = new HashMap<SampleDispatcher, SampleDispatcher>();
  }

  private synchronized SampleDispatcher add(SampleDispatcher identity) {
    SampleDispatcher renamed = alias.get(identity);
    if (renamed == null) {
      renamed = registry.add(identity, rename.rename(identity.getName()));
      alias.put(identity, renamed);
    }
    return renamed;
  }

  private synchronized SampleDispatcher get(SampleDispatcher identity) {
    return alias.get(identity);
  }

  public void register(SampleDispatcher identity) throws IOException {
    super.register(add(identity));
  }
  
  public void registerSampleRef(SampleDispatcher identity) throws IOException {
    super.registerSampleRef(add(identity));
  }
  
  public void begin(SampleDispatcher identity) throws IOException {
    super.begin(get(identity));
  }
  
  public void end(SampleDispatcher identity) throws IOException {
    super.end(get(identity));
  }

} 
